integration-tests/api/__tests__/store/gift-cards.js

```diff
@@ -60,4 +66,138 @@ describe("/store/gift-cards", () => {
+        gift_cards: [{ code: "GC_1" }, { code: "GC_2" }],
+      })
+
+      const response = await api.post(`/store/carts/${cart.id}/complete`)
+      expect(response.status).toEqual(200)
+      const orderId = response.data.data.id
+
+      const transaction1 = await dbConnection.manager.find(
+        GiftCardTransaction,
+        {
+          order_id: orderId,
+        }
+      )
+
+      expect(transaction1.length).toEqual(1)
+      expect(transaction1[0].amount).toEqual(100)
+    })
+  })
+
+  describe("gift card transaction selections", () => {
+    afterEach(async () => {
+      const db = useDb()
+      await db.teardown()
+    })
+
+    it("apply gift cards in ordered fashion, first by end_date, then by remaining balance", async () => {
+      const api = useApi()
+
+      const region = await simpleRegionFactory(dbConnection, {
+        currency_code: "usd",
+      })
+      const cart = await simpleCartFactory(dbConnection, {
+        region: region.id,
+      })
+      const product = await simpleProductFactory(dbConnection)
+      const ends_first = new Date(2050, 1, 1)
+      const ends_later = new Date(2050, 1, 2)
+      const giftCard1 = await simpleGiftCardFactory(dbConnection, {
+        id: "GC_1",
+        code: "GC_1",
+        region_id: region.id,
+        balance: 100,
+        value: 100,
+        ends_at: ends_later
+      })
+      const giftCard2 = await simpleGiftCardFactory(dbConnection, {
+        id: "GC_2",
+        code: "GC_2",
+        region_id: region.id,
+        balance: 200,
+        value: 200,
+        ends_at: ends_first
+      })
+      const giftCard3 = await simpleGiftCardFactory(dbConnection, {
+        id: "GC_3",
+        code: "GC_3",
+        region_id: region.id,
+        balance: 100,
+        value: 100,
+        ends_at: ends_first
+      })
+      const giftCard4 = await simpleGiftCardFactory(dbConnection, {
+        id: "GC_4",
+        code: "GC_4",
+        region_id: region.id,
+        balance: 100,
+        value: 100,
+      })
+
+      await api.post(`/store/carts/${cart.id}/line-items`, {
+        variant_id: product.variants[0].id,
+        quantity: 5,
+      })
+
+      await api.post(`/store/carts/${cart.id}`, {
+        email: "some@customer.com",
+        gift_cards: [{ code: "GC_1" }, { code: "GC_2" }, { code: "GC_3" }, { code: "GC_4" }],
+      })
+
+      const response = await api.post(`/store/carts/${cart.id}/complete`)
+      expect(response.status).toEqual(200)
+      const orderId = response.data.data.id
+
+      const transaction1 = await dbConnection.manager.find(
+        GiftCardTransaction,
+        {
+          order_id: orderId,
+        }
+      )
+
+      expect(transaction1.length).toEqual(4)
+      expect(transaction1[0].gift_card_id).toEqual(giftCard3.id)
+      expect(transaction1[0].gift_card_id).toEqual(giftCard2.id)
+      expect(transaction1[0].gift_card_id).toEqual(giftCard1.id)
+      expect(transaction1[0].gift_card_id).toEqual(giftCard4.id)
```

Comment:
thought: this is not correct you are checking if the same transaction has 4 different ids as its gift_card_id, which can't be true. More broadly, I don't think this test accurately captures the selection algorithm since the transactions are all created simultaneously so we can't really rely on the order of the transactions to check if they work correctly. 

The right approach as far as I can tell would be to do create multiple tests with the following:

**Test 1: `it("chooses gift cards that expire soon first")`**
1. Create n>2 gift cards with different end dates.
2. Construct the cart such that two gift cards will be needed to cover the cart amount. E.g., all gift cards have 100 balance and cart is a total of 200
3. Add all n gift cards to the cart.
4. Verify that
  a. Only two transactions are created.
  b. The two transactions are for the "soon" gift cards.


**Test 2: `it("chooses gift cards with low remaining balance first")`**
1. Create n>2 gift cards with different balances.
2. Construct the cart such that the two gift cards with the lowest balance can cover the cart.
3. Add all n gift cards to the cart.
4. Verify that
  a. Only two transactions are created.
  b. The two transactions are for the gift cards with lowest balances.
  
To be fair, these tests can't account for gift cards being ordered correctly by chance, but the probability of that happening would decrease with n so something like ~5 gift cards is probably good.

--

All that said - including the test for the sort order is not necessary for us to merge so perfectly fine to leave them out and just stick with the one that checks for the number of transactions created :)

Reply:
@srindom thanks

Annoying oversight on my behalf, not changing the indexes in the transactions array - and good point about needing to split them in two, due to not being sure that they will be created (and returned) in ordered fashion.

I also see an error in the run that GiftCardTransaction was not in the require list. Fixed that too.

Since I got going on it, and it's always good to learn, I've fixed the tests anyhow. But if you want to leave them out it's not a problem with me.